File: cache/vk/src/main/kotlin/Mem.kt
import verik.base.*
import verik.collection.*
import verik.data.*

class Mem: Module() {

    @input var clk = t_Boolean()
    @inout val bp  = t_MemRxBusPort()

    var mem = t_Array(exp(ADDR_WIDTH), t_UbitData())

    @seq fun update() {
        on (posedge(clk)) {
            bp.rsp_vld = false
            if (bp.rst) {
                for (i in range(exp(ADDR_WIDTH))) {
                    mem[i] = u(0)
                }
            } else {
                if (bp.req_op != Op.INVALID) {
                    println("mem received op=${bp.req_op} addr=0x${bp.req_addr} data=0x${bp.req_data}")
                    if (bp.req_op == Op.WRITE) {
                        mem[bp.req_addr] = bp.req_data
                    } else {
                        bp.rsp_data = mem[bp.req_addr]
                        bp.rsp_vld = true
                    }
                }
            }
        }
    }
}

File: cache/ch/src/main/scala/cache/Mem.scala
// See README.md for license details.

package cache

import chisel3._

class Mem extends Module {

  val io = IO(new Bundle {
    val rst = Input(Bool())
    val rx = new MemBundle()
  })

  val mem = RegInit(VecInit(Seq.fill(1 << Const.ADDR_WIDTH)(0.U(Const.DATA_WIDTH.W))))
  val data = Reg(UInt(Const.DATA_WIDTH.W))
  val valid = RegInit(false.B)

  io.rx.rspData := data
  io.rx.rspValid := valid

  data := 0.U
  valid := false.B

  when (io.rst) {
    mem.foreach { _ := 0.U }
  } .elsewhen (io.rx.reqOp =/= Op.invalid) {
    printf("mem received op=%d addr=0x%x data=0x%x\n", io.rx.reqOp, io.rx.reqAddr, io.rx.reqData)
    when (io.rx.reqOp === Op.write) {
      mem(io.rx.reqAddr) := io.rx.reqData
    } .otherwise {
      data := mem(io.rx.reqAddr)
      valid := true.B
    }
  }
}
